### DISCUSSION TITLE
Typescript error for `queryKey` when passing `Infinity` to `cacheTime`

### DISCUSSION THREAD

**Discussion by brenocastelo** [TanStack/query#5790] [Q&A]
Hi!

I'm implementing a data fetching where I want to keep the cache until the data is fetched again after being stale, so I set the `cacheTime` for `Infinity` (as described by the [docs](https://tanstack.com/query/v4/docs/react/reference/useQuery)). The issue is that when I set `cacheTime` to `Infinity`, typescript start complaining about the `queryKey`:

> No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ queryKey: string[]; queryFn: () => Promise<unknown>; staleTime: number; cacheTime: string; networkMode: string; }' is not assignable to parameter of type 'QueryKey'.
      Object literal may only specify known properties, and 'queryKey' does not exist in type 'readonly unknown[]'.
      
If I set `cacheTime` with a number, there's no issue. I notice that in the type definition for `cacheTime` is only a optional number, no sign of `Infinity`.

I would like to know if there is any way to handle this issue and why exactly, `queryKey` complains when we set 'cacheTime' value with a not number value.

Thanks inn advance.

**Comment by TkDodo** [comment]
you want `Infinity` (the number), not the string `"Infinity"`.

**Reply by TkDodo** [reply]
> and why exactly, queryKey complains

because we use function overloads in typescript, and TypeScript just shows errors for the last overload if nothing matches. This will get a lot better in v5 because we've removed all overloads: https://twitter.com/TkDodo/status/1683854950076411906

**Comment by mahcloud-tinker** [comment]
Hi, this still seems to be an issue in version 5.45.1.

This error seems to be resolved in version 5 by changing cacheTime to gcTime.
